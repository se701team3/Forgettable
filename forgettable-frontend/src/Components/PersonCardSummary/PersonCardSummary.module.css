@import '../../colours.css';

.PersonCardSummary {
    cursor: pointer;
    height: 120px;
    width: 240px;
    border-radius: var(--radius-card);
    background-color: var(--card);
    transition: 0.3s;
    display: flex;
    padding: 20px;
    box-sizing: border-box;
}

.ContentContainer {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.PersonCardSummary:hover{ 
    background-color: var(--hilit);
}

.TextInfoContainer {
    height: 100%;
    flex: 1;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: stretch;
    padding: 5px 0;
    align-items: flex-start;
    box-sizing: border-box;
}

.TextInfoContainer > h3 {
    font-size: var(--text-medium);
    font-weight: 600;
    margin: 0;
    text-align: left;
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: 1; /* number of lines to show */
            line-clamp: 1; 
    -webkit-box-orient: vertical;
}

.TextInfoContainer > p {
    font-size: var(--text-card-body-normal);
    margin: 0;
    color: var(--ltxt2);
    text-align: left;
    overflow: hidden;
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: 2; /* number of lines to show */
            line-clamp: 2; 
    -webkit-box-orient: vertical;
}
