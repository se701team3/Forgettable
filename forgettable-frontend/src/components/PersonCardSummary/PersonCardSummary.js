import {Avatar} from '@mui/material';
import React from 'react';
import classes from './PersonCardSummary.module.css';
import {getFirstMetTimeString} from '../../functions/dateFormatter';
import PropTypes from 'prop-types';
import {getImageSrcFromBuffer} from '../../functions/getImageSrcFromBuffer';

/*
 * Component for displaying information of a person in a list.
 * This is the small version of the person card.
 *
 * Author: Mercury Lin (lin8231)
 */
const PersonCardSummary = (props) => {
  const name = props.name.split(' ')[0];
  return (
    <div
      className={classes.PersonCardSummary}
      onClick={props.onClick}
      data-testid="container-card"
    >
      <div className={classes.ContentContainer}>
        <Avatar
          alt={name}
          // this style is written inline because MUI does not support className
          style={{
            height: '70px',
            width: '70px',
            marginRight: '14px',
            backgroundColor:
                getComputedStyle(document.body).getPropertyValue('--prmry'),
            fontSize:
                getComputedStyle(document.body)
                    .getPropertyValue('--text-xxlarge'),
          }}
          src={props.img ? getImageSrcFromBuffer(props.img) : props.img}
        />
        <div className={classes.TextInfoContainer}>
          <h3 data-testid="name-element">
            {name}
          </h3>
          <p data-testid="first-met-element">
              First met {getFirstMetTimeString(props.firstMet)}
          </p>
        </div>
      </div>
    </div>
  );
};

PersonCardSummary.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  img: PropTypes.string,
  firstMet: PropTypes.instanceOf(Date),
  onClick: PropTypes.func.isRequired,
};

export default PersonCardSummary;
