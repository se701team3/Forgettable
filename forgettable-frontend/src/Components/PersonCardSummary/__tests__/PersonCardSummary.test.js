import TestRenderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import {render, fireEvent, cleanup} from '@testing-library/react';
import {
  within,
  getByText,
  getByTestId,
} from '@testing-library/dom';
import PersonCardSummary from '../PersonCardSummary';

it('renders PersonCardSummary UI with correct hierarchy', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<PersonCardSummary
    name="Mercury Lin"
    onClick={() => {}}
  />);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});

afterEach(cleanup);

it('Name prop passed in is displayed correctly', () => {
  const name = 'Mercury Lin';
  const firstName = name.split(' ')[0];

  const {getByTestId} = render(
      <PersonCardSummary
        name={name}
        onClick={() => {}}
      />);

  const {getByText} = within(getByTestId('name-element'));

  expect(getByText(firstName)).toBeInTheDocument();
});

it('displays the correct format for firstMet for an existing date', () => {
  const twoYearsAgo = new Date(Date.now() - (365 * 24 * 60 * 60 * 1000 * 2));

  const {getByTestId} = render(
      <PersonCardSummary
        name="Mercury Lin"
        firstMet={twoYearsAgo}
        onClick={() => {}}
      />);

  const {getByText} = within(getByTestId('first-met-element'));

  expect(
      getByText('First met 2 years ago'),
  ).toBeInTheDocument();
});

it('displays the firstMet string when there is no date', () => {
  const {getByTestId} = render(
      <PersonCardSummary
        name="Mercury Lin"
        onClick={() => {}}
      />);

  const {getByText} = within(getByTestId('first-met-element'));

  expect(
      getByText('First met once upon a time'),
  ).toBeInTheDocument();
});

it('successfully fires event when card is clicked', () => {
  const handleClick = jest.fn();

  const {getByTestId} = render(
      <PersonCardSummary
        name=" Lin"
        onClick={handleClick}
      />);

  const node = getByTestId('container-card');

  fireEvent.click(node);
  expect(handleClick).toHaveBeenCalledTimes(1);
});


