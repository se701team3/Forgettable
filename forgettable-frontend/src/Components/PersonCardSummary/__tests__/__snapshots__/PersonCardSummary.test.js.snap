// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PersonCardSummary UI with correct hierarchy 1`] = `
<div
  className="PersonCardSummary"
  data-testid="container-card"
  onClick={[Function]}
>
  <div
    className="ContentContainer"
  >
    <ForwardRef(Avatar)
      alt="Mercury"
      style={
        Object {
          "backgroundColor": "",
          "fontSize": "",
          "height": "70px",
          "marginRight": "14px",
          "width": "70px",
        }
      }
    />
    <div
      className="TextInfoContainer"
    >
      <h3
        data-testid="name-element"
      >
        Mercury
      </h3>
      <p
        data-testid="first-met-element"
      >
        First met 
        once upon a time
      </p>
    </div>
  </div>
</div>
`;
